# Python Basics & Data Cleaning with Pandas

A hands-on exercise using the **Tableau *Sample – Superstore*** retail dataset.

**What we'll do (7 steps):**
1. Load a CSV dataset into a Pandas DataFrame
2. Explore the data (head / tail, shape, columns, data types)
3. Handle missing values (identify, fill / drop)
4. Basic operations (filter rows, select columns)
5. Remove duplicates
6. Create a derived column (`total_amount = price * quantity`)
7. Save the cleaned dataset as a new CSV

## Step 1 — Load a CSV into a DataFrame

In [24]:
import pandas as pd
import numpy as np

df = pd.read_csv("sample.csv", encoding="cp1252")
df.head()

,Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country,City,...,Postal Code,Region,Product ID,Category,Sub-Category,Product Name,Sales,Quantity,Discount,Profit
0,1,CA-2016-152156,11/8/2016,11/11/2016,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,...,42420,South,FUR-BO-10001798,Furniture,Bookcases,Bush Somerset Collection Bookcase,261.9600,2,0.00,41.9136
1,2,CA-2016-152156,11/8/2016,11/11/2016,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,...,42420,South,FUR-CH-10000454,Furniture,Chairs,"Hon Deluxe Fabric Upholstered Stacking Chairs,...",731.9400,3,0.00,219.5820
2,3,CA-2016-138688,6/12/2016,6/16/2016,Second Class,DV-13045,Darrin Van Huff,Corporate,United States,Los Angeles,...,90036,West,OFF-LA-10000240,Office Supplies,Labels,Self-Adhesive Address Labels for Typewriters b...,14.6200,2,0.00,6.8714
3,4,US-2015-108966,10/11/2015,10/18/2015,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,...,33311,South,FUR-TA-10000577,Furniture,Tables,Bretford CR4500 Series Slim Rectangular Table,957.5775,5,0.45,-383.0310
4,5,US-2015-108966,10/11/2015,10/18/2015,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,...,33311,South,OFF-ST-10000760,Office Supplies,Storage,Eldon Fold 'N Roll Cart System,22.3680,2,0.20,2.5164


## Step 2 — Explore the data

In [25]:
# Number of rows (observations) and columns (features)
df.shape

(9994, 21)

In [26]:
# Last 5 rows — the data can look different at the end than at the top
df.tail()

,Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country,City,...,Postal Code,Region,Product ID,Category,Sub-Category,Product Name,Sales,Quantity,Discount,Profit
9989,9990,CA-2014-110422,1/21/2014,1/23/2014,Second Class,TB-21400,Tom Boeckenhauer,Consumer,United States,Miami,...,33180,South,FUR-FU-10001889,Furniture,Furnishings,Ultra Door Pull Handle,25.248,3,0.2,4.1028
9990,9991,CA-2017-121258,2/26/2017,3/3/2017,Standard Class,DB-13060,Dave Brooks,Consumer,United States,Costa Mesa,...,92627,West,FUR-FU-10000747,Furniture,Furnishings,Tenex B1-RE Series Chair Mats for Low Pile Car...,91.960,2,0.0,15.6332
9991,9992,CA-2017-121258,2/26/2017,3/3/2017,Standard Class,DB-13060,Dave Brooks,Consumer,United States,Costa Mesa,...,92627,West,TEC-PH-10003645,Technology,Phones,Aastra 57i VoIP phone,258.576,2,0.2,19.3932
9992,9993,CA-2017-121258,2/26/2017,3/3/2017,Standard Class,DB-13060,Dave Brooks,Consumer,United States,Costa Mesa,...,92627,West,OFF-PA-10004041,Office Supplies,Paper,"It's Hot Message Books with Stickers, 2 3/4"" x 5""",29.600,4,0.0,13.3200
9993,9994,CA-2017-119914,5/4/2017,5/9/2017,Second Class,CC-12220,Chris Cortes,Consumer,United States,Westminster,...,92683,West,OFF-AP-10002684,Office Supplies,Appliances,"Acco 7-Outlet Masterpiece Power Center, Wihtou...",243.160,2,0.0,72.9480


In [27]:
# All column names
df.columns

Index(['Row ID', 'Order ID', 'Order Date', 'Ship Date', 'Ship Mode',
       'Customer ID', 'Customer Name', 'Segment', 'Country', 'City', 'State',
       'Postal Code', 'Region', 'Product ID', 'Category', 'Sub-Category',
       'Product Name', 'Sales', 'Quantity', 'Discount', 'Profit'],
      dtype='object')

In [28]:
# Data type of each column (int64 = whole numbers, float64 = decimals, object = text)
df.dtypes

Row ID             int64
Order ID          object
Order Date        object
Ship Date         object
Ship Mode         object
Customer ID       object
Customer Name     object
Segment           object
Country           object
City              object
State             object
Postal Code        int64
Region            object
Product ID        object
Category          object
Sub-Category      object
Product Name      object
Sales            float64
Quantity           int64
Discount         float64
Profit           float64
dtype: object

In [29]:
# One-line summary: size, dtypes, non-null counts, memory usage
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9994 entries, 0 to 9993
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Row ID         9994 non-null   int64  
 1   Order ID       9994 non-null   object 
 2   Order Date     9994 non-null   object 
 3   Ship Date      9994 non-null   object 
 4   Ship Mode      9994 non-null   object 
 5   Customer ID    9994 non-null   object 
 6   Customer Name  9994 non-null   object 
 7   Segment        9994 non-null   object 
 8   Country        9994 non-null   object 
 9   City           9994 non-null   object 
 10  State          9994 non-null   object 
 11  Postal Code    9994 non-null   int64  
 12  Region         9994 non-null   object 
 13  Product ID     9994 non-null   object 
 14  Category       9994 non-null   object 
 15  Sub-Category   9994 non-null   object 
 16  Product Name   9994 non-null   object 
 17  Sales          9994 non-null   float64
 18  Quantity

In [30]:
# Numeric summary statistics: count, mean, min, max, quartiles...
df.describe()

,Row ID,Postal Code,Sales,Quantity,Discount,Profit
count,9994.000000,9994.000000,9994.000000,9994.000000,9994.000000,9994.000000
mean,4997.500000,55190.379428,229.858001,3.789574,0.156203,28.656896
std,2885.163629,32063.693350,623.245101,2.225110,0.206452,234.260108
min,1.000000,1040.000000,0.444000,1.000000,0.000000,-6599.978000
25%,2499.250000,23223.000000,17.280000,2.000000,0.000000,1.728750
50%,4997.500000,56430.500000,54.490000,3.000000,0.200000,8.666500
75%,7495.750000,90008.000000,209.940000,5.000000,0.200000,29.364000
max,9994.000000,99301.000000,22638.480000,14.000000,0.800000,8399.976000


## Step 3 — Handle missing values

### 3a) Identify missing values

In [31]:
# How many missing values per column?
df.isnull().sum()

Row ID           0
Order ID         0
Order Date       0
Ship Date        0
Ship Mode        0
Customer ID      0
Customer Name    0
Segment          0
Country          0
City             0
State            0
Postal Code      0
Region           0
Product ID       0
Category         0
Sub-Category     0
Product Name     0
Sales            0
Quantity         0
Discount         0
Profit           0
dtype: int64

In [32]:
# What percentage of the table is missing in each column?
df.isnull().sum() / len(df) * 100

Row ID           0.0
Order ID         0.0
Order Date       0.0
Ship Date        0.0
Ship Mode        0.0
Customer ID      0.0
Customer Name    0.0
Segment          0.0
Country          0.0
City             0.0
State            0.0
Postal Code      0.0
Region           0.0
Product ID       0.0
Category         0.0
Sub-Category     0.0
Product Name     0.0
Sales            0.0
Quantity         0.0
Discount         0.0
Profit           0.0
dtype: float64

### 3b) Fill them — the strategy depends on the column

* **`Postal Code`** (an identifier) → fill with the *most common* value (`mode`)
* **`Region`** (categorical text) → fill with the *most common* value (`mode`)
* **`Sales`** (numbers) → fill with the *median* (resistant to outliers)

In [33]:
df["Postal Code"] = df["Postal Code"].fillna(df["Postal Code"].mode()[0])
df["Region"]      = df["Region"].fillna(df["Region"].mode()[0])
df["Sales"]       = df["Sales"].fillna(df["Sales"].median())

# Confirm: every column now has 0 missing values
df.isnull().sum()

Row ID           0
Order ID         0
Order Date       0
Ship Date        0
Ship Mode        0
Customer ID      0
Customer Name    0
Segment          0
Country          0
City             0
State            0
Postal Code      0
Region           0
Product ID       0
Category         0
Sub-Category     0
Product Name     0
Sales            0
Quantity         0
Discount         0
Profit           0
dtype: int64

### 3c) The alternative — drop them

Sometimes filling makes no sense (too much missing, or a value that genuinely can't
be guessed). Then drop the offending rows with `dropna()`. We already filled
everything, so we just preview what that would look like on a copy.

In [34]:
# What WOULD happen if we dropped every row with any missing value?
df.dropna().shape

(9994, 21)

## Step 4 — Basic operations: filter rows & select columns

### 4a) Filter rows with a condition


In [35]:
# Orders over $1,000
big_orders = df[df["Sales"] > 1000]
big_orders.shape

(468, 21)

In [36]:
# Combine conditions with & (and) — wrap each condition in parentheses
consumers_west = df[(df["Segment"] == "Consumer") & (df["Region"] == "West")]
consumers_west.shape

(1672, 21)

### 4b) Select columns


In [37]:
df[["Order ID", "Product Name", "Sales", "Quantity"]].head()

,Order ID,Product Name,Sales,Quantity
0,CA-2016-152156,Bush Somerset Collection Bookcase,261.9600,2
1,CA-2016-152156,"Hon Deluxe Fabric Upholstered Stacking Chairs,...",731.9400,3
2,CA-2016-138688,Self-Adhesive Address Labels for Typewriters b...,14.6200,2
3,US-2015-108966,Bretford CR4500 Series Slim Rectangular Table,957.5775,5
4,US-2015-108966,Eldon Fold 'N Roll Cart System,22.3680,2


In [38]:
# .loc: rows 0-4, columns "Order ID" and "Sales"
df.loc[0:4, ["Order ID", "Sales"]]

,Order ID,Sales
0,CA-2016-152156,261.9600
1,CA-2016-152156,731.9400
2,CA-2016-138688,14.6200
3,US-2015-108966,957.5775
4,US-2015-108966,22.3680


In [39]:
# .iloc: rows 0-4, column positions 1 (Order ID) and 17 (Sales)
df.iloc[0:4, [1, 17]]

,Order ID,Sales
0,CA-2016-152156,261.9600
1,CA-2016-152156,731.9400
2,CA-2016-138688,14.6200
3,US-2015-108966,957.5775


## Step 5 — Remove duplicates


In [40]:
# How many duplicate rows are hiding in the data?
df.duplicated().sum()

0

In [41]:
df = df.drop_duplicates()
print("Rows after removing duplicates:", len(df))

Rows after removing duplicates: 9994


## Step 6 — Create a derived column


In [42]:
df["Unit_Price"] = df["Sales"] / df["Quantity"]
df["total_amount"] = df["Unit_Price"] * df["Quantity"]

df[["Sales", "Quantity", "Unit_Price", "total_amount"]].head()

,Sales,Quantity,Unit_Price,total_amount
0,261.9600,2,130.9800,261.9600
1,731.9400,3,243.9800,731.9400
2,14.6200,2,7.3100,14.6200
3,957.5775,5,191.5155,957.5775
4,22.3680,2,11.1840,22.3680


In [43]:
# Sanity check: within 1 cent of Sales, for every row?
diff = (df["total_amount"] - df["Sales"]).abs()
print("Largest difference from Sales:", diff.max())
(diff < 0.01).all()

Largest difference from Sales: 4.547473508864641e-13


True

## Step 7 — Save the cleaned dataset

In [46]:
df.to_csv("superstore_cleaned.csv", index=False)
print("Saved superstore_cleaned.csv")

Saved superstore_cleaned.csv


In [48]:
# Verify the saved file loads back cleanly
cleaned = pd.read_csv("superstore_cleaned.csv")
print("Loaded back:", cleaned.shape)
print("Missing values:", cleaned.isnull().sum().sum())
print("Duplicate rows:", cleaned.duplicated().sum())

Loaded back: (9994, 23)
Missing values: 0
Duplicate rows: 0


## ✅ Summary

| Step | What we did | Before → After |
|---|---|---|
| 1. Load | `pd.read_csv("superstore_dirty.csv")` | 10,004 rows × 21 cols |
| 2. Explore | `head/tail/shape/columns/dtypes/info/describe` | understood the data |
| 3. Missing values | identified, filled (`mode`/`median`), `dropna` shown | 295 missing → 0 |
| 4. Basic ops | filtered rows, selected columns | — |
| 5. Duplicates | `df.drop_duplicates()` | 10 → 0 |
| 6. Derived col | `Unit_Price`, `total_amount = price × quantity` | +2 columns |
| 7. Save | `df.to_csv(..., index=False)` | `superstore_cleaned.csv` |

**Final cleaned dataset:** 9,994 rows × 23 columns — no missing values, no duplicates.

> **What the data said (a bonus peek):** `Sales` ranges from a tiny $0.44 order to a
> ~$22,638 order; median order ≈ $54.40. `Consumer` is the largest segment (5,191
> orders), and most orders come from the `West`, `East`, and `Central` US regions.